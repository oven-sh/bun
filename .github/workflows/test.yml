name: Test

concurrency:
  group: ${{ github.workflow }}-${{ inputs.run-id || github.run_id }}
  cancel-in-progress: true

on:
  push:
  workflow_call:
    inputs:
      run-id:
        type: string
        required: true

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/download-artifact@v4
        with:
          run-id: ${{ inputs.run-id || '8623134412' }}
          pattern: bun-*
      - run: |
          ls -la
