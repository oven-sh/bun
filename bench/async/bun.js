import { bench, run } from "mitata";

bench("sync", () => {});
bench("async", async () => {});
bench("await 1", async () => await 1);

await run();
