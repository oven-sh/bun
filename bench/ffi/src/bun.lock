{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@napi-rs/cli": "^2.10.1",
        "@node-rs/helper": "^1.3.3",
      },
    },
  },
  "packages": {
    "@napi-rs/cli": ["@napi-rs/cli@2.14.8", "", { "bin": { "napi": "scripts/index.js" } }, "sha512-IvA3s8BqohMdUbOkFn7+23u1dhIJZCkA8Xps7DD4SLdCMbcbUF6MUuKiqxuqmVHBFTaxU25sU56WdX3efqGgPw=="],

    "@napi-rs/triples": ["@napi-rs/triples@1.1.0", "", {}, "sha512-XQr74QaLeMiqhStEhLn1im9EOMnkypp7MZOwQhGzqp2Weu5eQJbpPxWxixxlYRKWPOmJjsk6qYfYH9kq43yc2w=="],

    "@node-rs/helper": ["@node-rs/helper@1.3.3", "", { "dependencies": { "@napi-rs/triples": "^1.1.0" } }, "sha512-p4OdfQObGN9YFy5WZaGwlPYICQSe7xZYyXB0sxREmvj1HzGKp5bPg2PlfgfMZEfnjIA882B9ZrnagYzZihIwjA=="],
  }
}
