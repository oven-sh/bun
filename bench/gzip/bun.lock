{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@babel/standalone": "7.24.10",
      },
    },
  },
  "packages": {
    "@babel/standalone": ["@babel/standalone@7.24.10", "", {}, "sha512-nGC37EKfmelpyCXto1pw6SBkD5ZQRdMbL6WISi28xWit9dtiy9dChU1WgEfzturUTxrmOGkMDRrCydFMA7uOaQ=="],
  }
}
