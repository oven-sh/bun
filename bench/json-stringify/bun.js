import { bench, run } from "mitata";

bench("JSON.stringify({hello: 'world'})", () =>
  JSON.stringify({ hello: "world" }),
);

const otherUint8Array = new Uint8Array(1024);
bench("Uint8Array.from(otherUint8Array)", () =>
  Uint8Array.from(otherUint8Array),
);

run();
