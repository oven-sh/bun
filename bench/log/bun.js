import { bench, run } from "mitata";

bench("console.log('hello')", () => console.log("hello"));
bench("console.log({ hello: 'object' })", () => console.log({ hello: "object" }));
await run();
