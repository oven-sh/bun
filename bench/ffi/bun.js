import { ptr, dlopen, CString, toBuffer } from "bun:ffi";
import { run, bench, group } from "mitata";

const { napiNoop, napiHash, napiString } = require(import.meta.dir + "/src/ffi_napi_bench.node");

const {
  symbols: {
    ffi_noop: { native: ffi_noop },
    ffi_hash: { native: ffi_hash },
    ffi_string: { native: ffi_string },
  },
} = dlopen(import.meta.dir + "/src/ffi_napi_bench.node", {
  ffi_noop: { args: [], returns: "void" },
  ffi_string: { args: [], returns: "ptr" },
  ffi_hash: { args: ["ptr", "u32"], returns: "u32" },
});

const bytes = new Uint8Array(64);

group("bun:ffi", () => {
  bench("noop", () => ffi_noop());
  bench("hash", () => ffi_hash(ptr(bytes), bytes.byteLength));

  bench("c string", () => new CString(ffi_string()));
});

if (process.env.SHOW_NAPI)
  group("bun:napi", () => {
    bench("noop", () => napiNoop());
    bench("hash", () => napiHash(bytes));

    bench("string", () => napiString());
  });

await run();
