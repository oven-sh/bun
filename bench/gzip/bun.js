import { run, bench, group } from "mitata";
import { gzipSync, gunzipSync } from "bun";

const data = new TextEncoder().encode("Hello World!".repeat(9999));

const compressed = gzipSync(data);

const libraries = ["zlib"];
if (Bun.semver.satisfies(Bun.version, ">=1.1.21")) {
  libraries.push("libdeflate");
}

const options = { library: undefined };
const benchFn = (name, fn) => {
  if (libraries.length > 1) {
    group(name, () => {
      for (const library of libraries) {
        bench(library, () => {
          options.library = library;
          fn();
        });
      }
    });
  } else {
    bench(name, () => {
      fn();
    });
  }
};

benchFn(`roundtrip - "Hello World!".repeat(9999))`, () => {
  gunzipSync(gzipSync(data, options), options);
});

benchFn(`gzipSync("Hello World!".repeat(9999)))`, () => {
  gzipSync(data, options);
});

benchFn(`gunzipSync("Hello World!".repeat(9999)))`, () => {
  gunzipSync(compressed, options);
});

await run();
