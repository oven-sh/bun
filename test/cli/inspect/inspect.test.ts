import { test, expect, afterEach } from "bun:test";
import { bunExe, bunEnv, randomPort } from "harness";
import { Subprocess, spawn } from "bun";
import { WebSocket } from "ws";

let inspectee: Subprocess;

const anyPort = expect.stringMatching(/^\d+$/);
const anyPathname = expect.stringMatching(/^\/[a-z0-9]+$/);
const tests = [
  {
    args: ["--inspect"],
    url: {
      protocol: "ws:",
      hostname: "localhost",
      port: "6499",
      pathname: anyPathname,
    },
  },
  {
    args: ["--inspect=0"],
    url: {
      protocol: "ws:",
      hostname: "localhost",
      port: anyPort,
      pathname: anyPathname,
    },
  },
  {
    args: [`--inspect=${randomPort()}`],
    url: {
      protocol: "ws:",
      hostname: "localhost",
      port: anyPort,
      pathname: anyPathname,
    },
  },
  {
    args: ["--inspect=localhost"],
    url: {
      protocol: "ws:",
      hostname: "localhost",
      port: "6499",
      pathname: anyPathname,
    },
  },
  {
    args: ["--inspect=localhost/"],
    url: {
      protocol: "ws:",
      hostname: "localhost",
      port: "6499",
      pathname: "/",
    },
  },
  {
    args: ["--inspect=localhost:0"],
    url: {
      protocol: "ws:",
      hostname: "localhost",
      port: anyPort,
      pathname: anyPathname,
    },
  },
  {
    args: ["--inspect=localhost:0/"],
    url: {
      protocol: "ws:",
      hostname: "localhost",
      port: anyPort,
      pathname: "/",
    },
  },
  {
    args: ["--inspect=localhost/foo/bar"],
    url: {
      protocol: "ws:",
      hostname: "localhost",
      port: "6499",
      pathname: "/foo/bar",
    },
  },
  {
    args: ["--inspect=127.0.0.1"],
    url: {
      protocol: "ws:",
      hostname: "127.0.0.1",
      port: "6499",
      pathname: anyPathname,
    },
  },
  {
    args: ["--inspect=127.0.0.1/"],
    url: {
      protocol: "ws:",
      hostname: "127.0.0.1",
      port: "6499",
      pathname: "/",
    },
  },
  {
    args: ["--inspect=127.0.0.1:0/"],
    url: {
      protocol: "ws:",
      hostname: "127.0.0.1",
      port: anyPort,
      pathname: "/",
    },
  },
  {
    args: ["--inspect=[::1]"],
    url: {
      protocol: "ws:",
      hostname: "[::1]",
      port: "6499",
      pathname: anyPathname,
    },
  },
  {
    args: ["--inspect=[::1]:0"],
    url: {
      protocol: "ws:",
      hostname: "[::1]",
      port: anyPort,
      pathname: anyPathname,
    },
  },
  {
    args: ["--inspect=[::1]:0/"],
    url: {
      protocol: "ws:",
      hostname: "[::1]",
      port: anyPort,
      pathname: "/",
    },
  },
  {
    args: ["--inspect=/"],
    url: {
      protocol: "ws:",
      hostname: "localhost",
      port: "6499",
      pathname: "/",
    },
  },
  {
    args: ["--inspect=/foo"],
    url: {
      protocol: "ws:",
      hostname: "localhost",
      port: "6499",
      pathname: "/foo",
    },
  },
  {
    args: ["--inspect=/foo/baz/"],
    url: {
      protocol: "ws:",
      hostname: "localhost",
      port: "6499",
      pathname: "/foo/baz/",
    },
  },
  {
    args: ["--inspect=:0"],
    url: {
      protocol: "ws:",
      hostname: "localhost",
      port: anyPort,
      pathname: anyPathname,
    },
  },
  {
    args: ["--inspect=:0/"],
    url: {
      protocol: "ws:",
      hostname: "localhost",
      port: anyPort,
      pathname: "/",
    },
  },
  {
    args: ["--inspect=ws://localhost/"],
    url: {
      protocol: "ws:",
      hostname: "localhost",
      port: anyPort,
      pathname: "/",
    },
  },
  {
    args: ["--inspect=ws://localhost:0/"],
    url: {
      protocol: "ws:",
      hostname: "localhost",
      port: anyPort,
      pathname: "/",
    },
  },
  {
    args: ["--inspect=ws://localhost:6499/foo/bar"],
    url: {
      protocol: "ws:",
      hostname: "localhost",
      port: "6499",
      pathname: "/foo/bar",
    },
  },
];

for (const { args, url: expected } of tests) {
  test(`bun ${args.join(" ")}`, async () => {
    inspectee = spawn({
      cwd: import.meta.dir,
      cmd: [bunExe(), ...args, "inspectee.js"],
      env: bunEnv,
      stdout: "ignore",
      stderr: "pipe",
    });

    let url: URL | undefined;
    let stderr = "";
    const decoder = new TextDecoder();
    for await (const chunk of inspectee.stderr as ReadableStream) {
      stderr += decoder.decode(chunk);
      for (const line of stderr.split("\n")) {
        try {
          url = new URL(line);
        } catch {
          // Ignore
        }
        if (url?.protocol.includes("ws")) {
          break;
        }
      }
      if (stderr.includes("Listening:")) {
        break;
      }
    }

    if (!url) {
      process.stderr.write(stderr);
      throw new Error("Unable to find listening URL");
    }

    const { protocol, hostname, port, pathname } = url;
    expect({
      protocol,
      hostname,
      port,
      pathname,
    }).toMatchObject(expected);

    const webSocket = new WebSocket(url);
    expect(
      new Promise<void>((resolve, reject) => {
        webSocket.addEventListener("open", () => resolve());
        webSocket.addEventListener("error", cause => reject(new Error("WebSocket error", { cause })));
        webSocket.addEventListener("close", cause => reject(new Error("WebSocket closed", { cause })));
      }),
    ).resolves.toBeUndefined();

    webSocket.send(JSON.stringify({ id: 1, method: "Runtime.evaluate", params: { expression: "1 + 1" } }));
    expect(
      new Promise(resolve => {
        webSocket.addEventListener("message", ({ data }) => {
          resolve(JSON.parse(data.toString()));
        });
      }),
    ).resolves.toMatchObject({
      id: 1,
      result: {
        result: {
          type: "number",
          value: 2,
        },
      },
    });

    webSocket.close();
  });
}

// FIXME: Depends on https://github.com/oven-sh/bun/pull/4649
test.todo("bun --inspect=ws+unix:///tmp/inspect.sock");

afterEach(() => {
  inspectee?.kill();
});
