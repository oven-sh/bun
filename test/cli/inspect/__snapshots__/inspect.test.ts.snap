// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`junit reporter 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<testsuites>
  <testsuite name="<dir>/a.test.js" tests="2" failures="1" errors="0" skipped="0" timestamp="2024-12-17T15:37:38.935Z">
    <testcase classname="<dir>/a.test.js" name="fail"       <failure message="Test failed" type="AssertionError">
        Error: expect(received).toBe(expected)

Expected: 2
Received: 1


  at <dir>/a.test.js:4:19


      </failure>
    </testcase>
    <testcase classname="<dir>/a.test.js" name="success"     </testcase>
  </testsuite>
</testsuites>"
`;
